src/ilastik-packager.imjoy.html

```diff
@@ -0,0 +1,348 @@
+
+
+async function convertZenodoFileUrl(url) {
+  const myRegexp = /zenodo.org\/record\/([a-zA-Z0-9-]+)\/files\/(.*)/g;
+  const match = myRegexp.exec(url);
+  if (!match || match.length !== 3) {
+    return null
+  }
+  const [fullUrl, depositId, fileName] = match;
+  if(url.includes('sandbox.zenodo.org'))
+    url = `https://sandbox.zenodo.org/api/records/${depositId}`;
+  else
+    url = `https://zenodo.org/api/records/${depositId}`;
+  const blob = await fetch(url).then(r => r.blob());
+  const data = JSON.parse(await new Response(blob).text());
+  const fn = fileName.split("?")[0];
+  const fileObj = data.files.filter(file => {
+    return file.key === fn;
+  })[0];
+  return fileObj.links.self;
+}
+
+async function fetchFile(url, showMessage, showProgress){
+    url = await convertZenodoFileUrl(url) || url;
+    const response = await axios({
+    url,
+    method: 'GET',
+    responseType: 'blob',
+    onDownloadProgress: (progressEvent) => {
+      showMessage(`Downloading weights ${progressEvent.loaded}/${progressEvent.total}`);
+      if(progressEvent.total)
+      showProgress(progressEvent.loaded/progressEvent.total)
+    }
+  })
+  const filename = url.split('/').pop().split('#')[0].split('?')[0];
+  const blob = new Blob([response.data]);
+  const file = new File([blob], filename, {type: "application/octet-stream", lastModified: Date.now()});
+  return file
+}
+
+
+async function resolveUrl(rootUrl, url){
+    if(!url.startsWith('http')){
+       url = rootUrl + '/' + url.replace('./', '')
+    }
+    if(isZenodoFileUrl(url)) url = await convertZenodoFileUrl(url)
+    return url
+}
+
+function getFileName(url){
+  return './' + url.split('/').pop().split('#')[0].split('?')[0]
+}
+
+
+async function downloadModel(rootUrl, spec, weightsFormat, showMessage, showProgress){
+  const zip = new JSZip()
+  const exportSpec = JSON.parse(JSON.stringify(spec));
+
+  await showMessage('Downloading files...')
+  let files = []
+  const weightsURL = spec.weights[weightsFormat].source
+  if(weightsURL){
+    files.push(weightsURL)
+    exportSpec.weights[weightsFormat].source = getFileName(weightsURL)
+  }
+  if(spec.attachments && spec.attachments['files']){
+    const fs = spec.attachments['files']
+    files = files.concat(fs)
+    const exportedFs = exportSpec.attachments['files']
+    for(let i=0;i<fs.length;i++){
+      exportedFs[i] = getFileName(fs[i])
+    }
+  }
+  if(spec.weights[weightsFormat].attachments && spec.weights[weightsFormat].attachments.files){
+    const fs = spec.weights[weightsFormat].attachments.files
+    files = files.concat(fs)
+    const exportedFs = exportSpec.weights[weightsFormat].attachments.files
+    for(let i=0;i<fs.length;i++){
+      exportedFs[i] = getFileName(fs[i])
+    }
+  }
+  // deprecate this after v0.4.0
+  if(spec.source && spec.source.startsWith('./')){
+    const src = spec.source.split(':').slice(0,1) // e.g. ./mynetwork:UNet
+    files.push(src)
+    exportSpec.source = getFileName(src)
+  }
```

Comment:
@oeway  this fails for models with 
```
VM2262:79 Uncaught (in promise) TypeError: url.split is not a function
    at getFileName (eval at execute (VM2248 imjoy-rpc.min.js:1), <anonymous>:79:21)
    at downloadModel (eval at execute (VM2248 imjoy-rpc.min.js:1), <anonymous>:115:25)
    at async un.download (eval at execute (VM2248 imjoy-rpc.min.js:1), <anonymous>:231:11)
    ```
    slice() returns a list, but getFileName expects a string... 
    
    Because I'm really unfamiliar with javascript I would like to swtich to python (like this: https://gist.github.com/oeway/2d4b5899424a14d8e90ad908d4cec364/) 
    
    only problem I see there, how do I present a "Download" button? Any advice?

Reply:
I think I found it:
https://imjoy.io/docs/#/api?id=api-functions: `api.exportFile(file, name)`

great when documentation is all there!!! ❤️

Reply:
Have you solved this?

Reply:
writing the python packager first and then I'll make the bioengine app for it.... 
I'll make a separate PR when I have a draft (this one is the backup in javascript...

Reply:
As of now, I would not use the python packager directly to run from the website, mostly because the startup time with binder can be several minutes, which might be fine if we run a model, but won't be good for each download of models.

Reply:
I just made a commit to fix this issue, it's simple to fix, we just need to replace `.slice(0, 1)` to `[0]`.